### PR TITLE
add retry for failing spawn_pipeline

### LINKED ISSUE #463: Better supervisor
Right now the pipeline supervisor retry logic has two problems.

First, it just retries right away. We should probably have some type of exponential backoff...
Maybe we should log a warning on the first issue, and an error on the second to help users configure pagers?
(@guilload what is the right practise here?)

Second (which is way more important)
Right now, if an error happens during the pipeline creation, we just fail.
This should also be part of the respawn process. 

The reason why it is deactivated is because, for the current CLI experience, we want to exit if there is a connectivity problem. There is a consensus however that this should be solved in #701.

### PR DESCRIPTION
**Pull request by PSeitz**
fixes #463

### Description
add spawn command for supervisor with retry

### How was this PR tested?
Describe how you tested this PR.

### CODE REVIEW

File: quickwit-indexing/src/actors/pipeline_supervisor.rs [hunk 1]
```diff
@@ -405,16 +407,41 @@ impl IndexingPipelineSupervisor {
...
+                            retry_count: retry_count + 1,
+                        },
+                    );
+                }
+            }
+            _ => {
+                self.spawn_pipeline(ctx).await?;
+            }
+        };
+        Ok(())
```
Comment (fulmicoton) [inline quickwit-indexing/src/actors/pipeline_supervisor.rs:440]: This is too convoluted.

```rust
if let Err(spawn_error) = self.spawn_pipeline(ctx).await {
   if let Some(duration_before_retry) = wait_duration_before_retry(retry_count) {
      ctx.schedule_self_msg_blocking(
             duration,
             Msg::Spawn {
                 retry_count: retry_count + 1,
             },
     );
   } else {
      return Err(spawn_error);
   }
}
Ok(())
```

or with a functional style

```rust
self.spawn_pipeline(ctx).await
     .or_else(|err| {
          let duration_before_retry = wait_duration_before_retry(retry_count)
              .unwrap_or_else(|| Err(err))?;
           ctx.schedule_self_msg_blocking(
                    duration,
                    Msg::Spawn {
                        retry_count: retry_count + 1,
                    });
            });
            Ok(())
    })
```

File: quickwit-indexing/src/actors/pipeline_supervisor.rs [hunk 2]
```diff
@@ -492,14 +521,106 @@ mod tests {
...
+        metastore
+            .expect_mark_splits_for_deletion()
+            .returning(|_, _| Ok(()));
+
+        metastore
+            .expect_index_metadata()
+            .withf(|index_id| index_id == "test-index")
+            .returning(move |_| {
+                if num_fails != 0 {
+                    num_fails -= num_fails.saturating_sub(1);
```
Comment (fulmicoton) [inline quickwit-indexing/src/actors/pipeline_supervisor.rs:549]:
```suggestion
                    num_fails -= 1;
```

File: quickwit-indexing/src/actors/pipeline_supervisor.rs [hunk 3]
```diff
@@ -492,14 +521,106 @@ mod tests {
...
+                    Err(MetastoreError::ConnectionError {
+                        message: "MetastoreError Alarm".to_string(),
+                    })
+                } else {
```
Comment (fulmicoton) [inline quickwit-indexing/src/actors/pipeline_supervisor.rs:553]: Early return would be nicer here.

File: quickwit-indexing/src/actors/pipeline_supervisor.rs [hunk 4]
```diff
@@ -405,16 +407,41 @@ impl IndexingPipelineSupervisor {
...
 
+    async fn process_spawn(
+        &mut self,
+        ctx: &ActorContext<Msg>,
+        retry_count: usize,
+    ) -> Result<(), ActorExitStatus> {
+        match retry_count {
+            0 | 1 | 2 => {
+                let duration = if retry_count == 0 {
+                    Duration::from_millis(1000)
```
Comment (fulmicoton) [inline quickwit-indexing/src/actors/pipeline_supervisor.rs:418]: There is a Duration::from_secs.

File: quickwit-indexing/src/actors/pipeline_supervisor.rs [hunk 5]
```diff
@@ -405,16 +407,41 @@ impl IndexingPipelineSupervisor {
...
+                } else if retry_count == 1 {
+                    Duration::from_millis(2000)
+                } else {
+                    Duration::from_millis(5000)
+                };
+
+                if let Err(spawn_error) = self.spawn_pipeline(ctx).await {
+                    warn!(err=?spawn_error, "Error while spawn_pipeline, waiting 1000ms");
```
Comment (fulmicoton) [inline quickwit-indexing/src/actors/pipeline_supervisor.rs:426]: Warning is incorrect.

File: quickwit-indexing/src/actors/pipeline_supervisor.rs [hunk 6]
```diff
@@ -405,16 +407,41 @@ impl IndexingPipelineSupervisor {
...
+                    ctx.schedule_self_msg_blocking(
+                        duration,
+                        Msg::Spawn {
+                            retry_count: retry_count + 1,
+                        },
+                    );
+                }
+            }
+            _ => {
+                self.spawn_pipeline(ctx).await?;
```
Comment (fulmicoton) [inline quickwit-indexing/src/actors/pipeline_supervisor.rs:437]: I think actually we want to keep retrying forever, at least for the moment.

Soooo:
- retry forever
- exponential backoff until the retry goes up to say 5mn, after which we stagnate to 5mn. (It actually simpliy the code too)

**Review by fulmicoton — CHANGES_REQUESTED** [review]
See comments inline.

**Review by fulmicoton — CHANGES_REQUESTED** [review]
see comments.

File: quickwit-indexing/src/actors/pipeline_supervisor.rs (hunk 6 above)
Comment (PSeitz) [reply quickwit-indexing/src/actors/pipeline_supervisor.rs]: if it's just for now, I would keep the logic like it is and just always return Some(duration)

Comment (fulmicoton) [reply quickwit-indexing/src/actors/pipeline_supervisor.rs]: Makes sense.

**Comment by fulmicoton** [comment]
@PSeitz I think most of my comments were lost and left unaddressed. 
Should I readd them?

**Comment by PSeitz** [comment]
> @PSeitz I think most of my comments were lost and left unaddressed. Should I readd them?

Ah sorry, yeah please readd them

File: quickwit-indexing/src/actors/indexing_pipeline.rs [hunk 7]
```diff
@@ -419,19 +421,44 @@ impl IndexingPipeline {
...
 
+    async fn process_spawn(
+        &mut self,
+        ctx: &ActorContext<Msg>,
+        retry_count: usize,
+    ) -> Result<(), ActorExitStatus> {
+        match retry_count {
+            0 | 1 | 2 => {
+                let duration = if retry_count == 0 {
+                    Duration::from_millis(1000)
```
Comment (fulmicoton) [inline quickwit-indexing/src/actors/indexing_pipeline.rs:432]: from_secs exists.

File: quickwit-indexing/src/actors/indexing_pipeline.rs [hunk 8]
```diff
@@ -419,19 +421,44 @@ impl IndexingPipeline {
...
+                } else if retry_count == 1 {
+                    Duration::from_millis(2000)
```
Comment (fulmicoton) [inline quickwit-indexing/src/actors/indexing_pipeline.rs:434]: This is too convoluted.

```rust
if let Err(spawn_error) = self.spawn_pipeline(ctx).await {
   if let Some(duration_before_retry) = wait_duration_before_retry(retry_count) {
      ctx.schedule_self_msg_blocking(
             duration,
             Msg::Spawn {
                 retry_count: retry_count + 1,
             },
     );
   } else {
      return Err(spawn_error);
   }
}
Ok(())
```

or with a functional style

```rust
self.spawn_pipeline(ctx).await
     .or_else(|err| {
          let duration_before_retry = wait_duration_before_retry(retry_count)
              .unwrap_or_else(|| Err(err))?;
           ctx.schedule_self_msg_blocking(
                    duration,
                    Msg::Spawn {
                        retry_count: retry_count + 1,
                    });
            });
            Ok(())
    })
```

File: quickwit-indexing/src/actors/indexing_pipeline.rs [hunk 9]
```diff
@@ -550,6 +579,98 @@ mod tests {
...
+        metastore
+            .expect_mark_splits_for_deletion()
+            .returning(|_, _| Ok(()));
+
+        metastore
+            .expect_index_metadata()
+            .withf(|index_id| index_id == "test-index")
+            .returning(move |_| {
+                if num_fails != 0 {
+                    num_fails -= num_fails.saturating_sub(1);
```
Comment (fulmicoton) [inline quickwit-indexing/src/actors/indexing_pipeline.rs:599]: !?

Comment (fulmicoton) [reply quickwit-indexing/src/actors/indexing_pipeline.rs:599]:
```suggestion
                    num_fails -= 1;
```

**Review by fulmicoton — CHANGES_REQUESTED** [review]
See comments.

File: quickwit-indexing/src/actors/indexing_pipeline.rs [hunk 10]
```diff
@@ -419,19 +421,44 @@ impl IndexingPipeline {
         Ok(())
     }
 
+    /// retry_count, wait_time_sec
+    /// 1   1
+    /// 2   4
+    /// 3   8
+    /// 4   16
```
Comment (fulmicoton) [inline quickwit-indexing/src/actors/indexing_pipeline.rs:428]: /// ...
/// >= 9    5mn

Comment (fulmicoton) [inline quickwit-indexing/src/actors/indexing_pipeline.rs:428]:
```suggestion
    /// retry_count, wait_time
    /// 0   1s
    /// 1   2s
    /// 2   4s
    /// 3   8s
    /// ...
    /// >=8   5mn
```

File: quickwit-indexing/src/actors/indexing_pipeline.rs [hunk 11]
```diff
@@ -419,19 +421,44 @@ impl IndexingPipeline {
...
+    fn wait_duration_before_retry(retry_count: usize) -> Option<Duration> {
```
Comment (fulmicoton) [inline quickwit-indexing/src/actors/indexing_pipeline.rs:429]: unit test on the mapping please.

File: quickwit-indexing/src/actors/indexing_pipeline.rs [hunk 12]
```diff
@@ -419,19 +421,44 @@ impl IndexingPipeline {
...
-        if self.handlers.is_none() {
-            // TODO Accept errors in spawning. See #463.
-            self.spawn_pipeline(ctx).await?;
-            // if let Err(spawn_error) = self.spawn_pipeline(ctx).await {
-            //    // only retry n-times.
-            //    error!(err=?spawn_error, "Error while spawning");
-            //    self.terminate().await;
-            // }
-        } else {
+        if self.handlers.is_some() {
```
Comment (fulmicoton) [inline quickwit-indexing/src/actors/indexing_pipeline.rs:461]: I am lost... When do we respawn if workers have failed now?

**Review by fulmicoton — CHANGES_REQUESTED** [review]
Question: When do we respawn if the worker fail?

File: quickwit-indexing/src/actors/indexing_pipeline.rs [hunk 13]
```diff
@@ -543,13 +574,94 @@ mod tests {
...
+            .returning(|_, _, _, _| Ok(Vec::new()));
+        metastore
+            .expect_mark_splits_for_deletion()
+            .returning(|_, _| Ok(()));
+
+        metastore
+            .expect_index_metadata()
+            .withf(|index_id| index_id == "test-index")
+            .returning(move |_| {
+                if num_fails == 0 {
```
Comment (fulmicoton) [inline quickwit-indexing/src/actors/indexing_pipeline.rs:600]: I wonder if it is possible to do that in mockall with "times".

Basically register the n failure first, and then register the success.

**Review by fulmicoton — CHANGES_REQUESTED** [review]
see comments

File: quickwit-indexing/src/actors/indexing_pipeline.rs (hunk 12 above)
Comment (PSeitz) [reply quickwit-indexing/src/actors/indexing_pipeline.rs:469]: I added it to the observer now

**Review by fulmicoton — APPROVED** [review]

**Pull request merged**